[clangd] Support outgoing calls in call hierarchy

HighCommander4 opened the pull request.

qchateau committed:
[clangd] Support outgoing calls in call hierarchy

The implementation is very close the the incoming
calls implementation. The results of the outgoing
calls are expected to be the exact symmetry of the
incoming calls.

Differential Revision: https://reviews.llvm.org/D93829

HighCommander4 committed:
Implement the RefKind::Call optimization

HighCommander4 committed:
Implement the simple lookup optimization

HighCommander4 committed:
Address other review comments

HighCommander4 commented:
(Not sure what happened here, the PR was somehow created while I was still writing the description. Anyways, updating the title now.)

HighCommander4 commented:
Since Phabricator has been [taken down](https://discourse.llvm.org/t/llvm-phabricator-turndown/76137), I'm resubmitting the patch implementing outgoing calls in call hierarchy that was previously posted at https://reviews.llvm.org/D93829.

Previous discussion can still be seen at the Phabricator link which now links to a static snapshot of the review.

HighCommander4 commented:
@sam-mccall review ping :)

I would particularly appreciate feedback on whether I should plan to set aside some time to implement the "deep lookup optimization" (from [this comment](https://reviews.llvm.org/D93829#4258101)), or whether the 2.5% increase in index memory usage (which I've achieved by implementing everything _except_ the deep lookup optimization from that comment) is acceptable.

pidgeon777 commented:
@sam-mccall also pinging to assist @HighCommander4 🙂

HighCommander4 commented:
Shopping around for some potential alternative reviewers :)

pidgeon777 commented:
I was thinking that it's a real pity this branch is so hard to get merged 😅 I think other than the review it would be 100% done? I mean, no big further modifications would be involved?

HighCommander4 commented:
> I think other than the review it would be 100% done? I mean, no big further modifications would be involved?

In terms of functionality, I'd say the patch is complete.

In terms of performance (and specifically memory usage), it's a matter of judgment of what we consider good enough. The current patch increases the memory usage of the index by 2.5% on the example workload of the LLVM codebase. That's not too bad (and a definite improvement over the 8.2% increase for the original version of the patch), but there is a possibility of doing a more involved rework of the index's in-memory representation (discussed under the name "deep lookup optimization" in the Phabricator thread) which would actually result in a net decrease in memory usage even with the feature.

One of the key pieces of feedback I'm hoping to get on this patch is whether the 2.5% memory usage increase is acceptable, or if we should consider the deep lookup optimization as a prerequisite.

HighCommander4 commented:
@kadircet perhaps you might be able to pick up this review?

Or, barring of a full review, your opinion on the directional question in [this comment](https://reviews.llvm.org/D93829#4654786) would be appreciated as well:

> how would you feel about proceeding with the patch in its current state, with the memory usage increase brought down from 8.2% to 2.5% thanks to the combination of the simple lookup optimization + RefKind filtering, and leaving the "deep lookup optimization" to be explored in a future change?

pidgeon777 commented:
Hopefully we're closer and closer to a merge, this really is a missing feature. I think this is the only LSP provider supporting just only one of the callHierarchy methods.

HighCommander4 commented:
@ckandeler, perhaps you would be willing to review this patch? I don't seem to have gotten any traction from my usual reviewers over the course of the past year.

ckandeler commented on clang-tools-extra/clangd/test/type-hierarchy.test:
```diff
@@ -88,56 +89,6 @@
 # CHECK-NEXT:     }
 # CHECK-NEXT:  ]
 ---
-{"jsonrpc":"2.0","id":2,"method":"typeHierarchy/subtypes","params":{"item":{"uri":"test:///main.cpp","data":{"parents":[{"parents":[{"parents":[],"symbolID":"FE546E7B648D69A7"}],"symbolID":"ECDC0C46D75120F4"}],"symbolID":"8A991335E4E67D08"},"name":"Child2","kind":23,"range":{"end":{"character":13,"line":3},"start":{"character":7,"line":3}},"selectionRange":{"end":{"character":13,"line":3},"start" ...
```
What was the reasoning for removing this entirely?

HighCommander4 replied:
Huh; I don't recall. The test case seems to still be valid, so I restored it.

ckandeler commented on clang-tools-extra/clangd/XRefs.cpp:
```diff
@@ -2346,11 +2346,11 @@ outgoingCalls(const CallHierarchyItem &Item, const SymbolIndex *Index) {
+                    Kind != SK::ClassMethod && Kind != SK::StaticMethod &&
+                    Kind != SK::Constructor && Kind != SK::Destructor &&
+                    Kind != SK::ConversionFunction);
+    assert(!NotCall);
```
Can we get rid of the double negation? It just seems to obfuscate the logic for no particular reason.

ckandeler commented on clang-tools-extra/clangd/XRefs.cpp:
```diff
@@ -2346,11 +2346,11 @@ outgoingCalls(const CallHierarchyItem &Item, const SymbolIndex *Index) {
     // Filter references to only keep function calls
```
This comment doesn't really apply anymore.

HighCommander4 replied:
Reworded

i-garrison commented on clang-tools-extra/clangd/XRefs.cpp:
```diff
@@ -1751,6 +1752,7 @@ static std::optional<HierarchyItem> symbolToHierarchyItem(const Symbol &S,
   }
   HierarchyItem HI;
   HI.name = std::string(S.Name);
+  HI.detail = (S.Scope + S.Name).str();
```
adding signature looks a bit more helpful, maybe one day add return type info too
```suggestion
  HI.detail = (S.Scope + S.Name + S.Signature).str();
```

HighCommander4 replied:
I agree, showing the signature here would be quite useful.

I actually started implementing this for incoming calls in https://github.com/clangd/clangd/issues/915, but ran into a complication (our current indexing process doesn't actually store the signature for all functions).

This is solvable with some indexer changes, but I'll defer to a fix for that issue, which will now (once this patch is merged) benefit both incoming and outgoing calls.

HighCommander4 committed:
Address newer review comments

HighCommander4 commented on clang-tools-extra/clangd/XRefs.cpp:
```diff
@@ -1700,6 +1700,7 @@ declToHierarchyItem(const NamedDecl &ND, llvm::StringRef TUPath) {
 
   HierarchyItem HI;
   HI.name = printName(Ctx, ND);
+  // FIXME: Populate HI.detail the way we do in symbolToHierarchyItem?
```
The difference between `declToHierarchyItem` and `symbolToHierarchyItem` is the former is used for symbols from the AST, and the latter for symbols from the index. They should both ideally produce the same result, but the implementation (and sometimes the amount of detail available in the respective inputs) is different.

i-garrison replied:
I see. Ideally there should be some recommended use model for `c++` which call hierarchy clients/visualisers (e.g. lsp4e) could implement to show expected function signatures. Like "obtain details field from call hierarchy item and use it as function signature" and then lsp4e could append details to name. We probably can later tune xrefs.cpp implementation to provide required details where needed. Note that function return value is still not available, and as far as I can see there is no place in protocol to put it in.

HighCommander4 commented:
Thanks @ckandeler for the review! I addressed the comments in https://github.com/llvm/llvm-project/pull/77556/commits/bad4f7264433417763ea35bc1104fb087c8725e5.

ckandeler reviewed (approved):
Appears to work fine.
Regarding the slight memory increase: If anyone feels strongly about it, they had plenty of time to chime in.

pidgeon777 commented:
First of all, I want to express my gratitude to all contributors working on the integration of changes and tests.

In today's development landscape, especially when working with AI-assisted coding, providing the best possible code context is crucial. The integration of outgoing calls functionality can significantly improve this aspect.

The main benefit of implementing outgoing calls would be context extension. This feature would:

- Allow developers to see function call relationships
- Help identify potential impacts when modifying functions
- Provide better context for AI-assisted development
- Improve code maintenance and refactoring workflows

When modifying a function, developers would have immediate visibility of all called functions that might need corresponding updates, reducing the risk of introducing bugs.

HighCommander4 commented:
> Appears to work fine. Regarding the slight memory increase: If anyone feels strongly about it, they had plenty of time to chime in.

Thanks! I'll go ahead and merge.

HighCommander4 commented:
Reverted inhttps://github.com/llvm/llvm-project/pull/117668 due to build breakage.

HighCommander4 commented:
> Reverted in #117668 due to build breakage.

The buildkite runs in this PR were green, but the PR was based on an older baseline, and some of the affected files have since changed in ways that require an update to the patch.

HighCommander4 commented:
Resubmitted with the build failures fixed at https://github.com/llvm/llvm-project/pull/117673.
